subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ProviderMigrationArchitectureTest.java

```diff
@@ -0,0 +1,168 @@
+import static com.tngtech.archunit.core.domain.JavaMember.Predicates.declaredIn;
+import static com.tngtech.archunit.core.domain.properties.CanBeAnnotated.Predicates.annotatedWith;
+import static com.tngtech.archunit.lang.conditions.ArchPredicates.are;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+import static org.gradle.architecture.test.ArchUnitFixture.freeze;
+import static org.gradle.architecture.test.ArchUnitFixture.public_api_methods;
+
+@AnalyzeClasses(packages = "org.gradle")
+public class ProviderMigrationArchitectureTest {
+    private static final DescribedPredicate<JavaMethod> getters = new DescribedPredicate<JavaMethod>("getters") {
+        @Override
+        public boolean apply(JavaMethod input) {
+            PropertyAccessorType accessorType = PropertyAccessorType.fromName(input.getName());
+            return accessorType == PropertyAccessorType.GET_GETTER || accessorType == PropertyAccessorType.IS_GETTER;
+        }
+    };
+
+    private static final DescribedPredicate<JavaClass> class_with_any_mutable_property = new DescribedPredicate<JavaClass>("class with any mutable property") {
+        @Override
+        public boolean apply(JavaClass input) {
+            return input.getAllMethods().stream()
+                .filter(getters::apply)
+                .anyMatch(ProviderMigrationArchitectureTest::hasSetter);
+        }
+    };
+
+    private static final DescribedPredicate<JavaMethod> mutable_public_API_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(not(declaredIn(assignableTo(Task.class))))
+        .and(not(declaredIn(StartParameter.class)))
+        .and(not(declaredIn(Configuration.class)))
+        .and(not(declaredIn(FileCollection.class)))
+        .and(not(declaredIn(ConfigurableFileCollection.class)))
+        .and(are(declaredIn(class_with_any_mutable_property)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .as("mutable public API properties");
+
+    @SuppressWarnings("deprecation")
+    private static final DescribedPredicate<JavaMethod> task_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(declaredIn(assignableTo(Task.class)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .and(not(declaredIn(Task.class)))
+        .and(not(declaredIn(DefaultTask.class)))
+        .and(not(declaredIn(org.gradle.api.internal.AbstractTask.class)))
+        .as("task properties");
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_be_providers = freeze(methods()
+        .that(are(mutable_public_API_properties))
+        .and().doNotHaveRawReturnType(TextResource.class)
+        .and().doNotHaveRawReturnType(assignableTo(FileCollection.class))
+        .should(haveProviderReturnType()));
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_be_configurable_file_collections = freeze(methods()
```

Comment:
should this be just _should_be_file_collections now that we handle mutable and immutable properties?

Reply:
Since we only look at methods which return `FileCollection`, this is essentially checking that the mutable ones return `ConfigurableFileCollection`. So I think the test name is good now.